services/scraper/search_keyword.go

```diff
@@ -0,0 +1,86 @@
+package scraper
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/beego/beego/v2/core/logs"
+	"github.com/gocolly/colly/v2"
+)
+
+type SearchKeywordService struct {
+	Keyword string
+
+	synchronous  bool
+	searchResult *searchKeywordResult
+}
+
+type searchKeywordResult struct {
+	Keyword  string
+	VisitURL string
+	NonAds   []string
+	OtherAds []string
+	TopAds   []string
+}
+
+var selectorList = map[string]string{
+	"nonAds":        "#search .g .yuRUbf > a",
+	"bottomLinkAds": "#tadsb .d5oMvf > a",
+	"otherAds":      "#rhs .pla-unit a.pla-unit-title-link",
+	"topImageAds":   "#tvcap .pla-unit a.pla-unit-title-link",
+	"topLinkAds":    "#tads .d5oMvf > a",
+}
+
+const collectingURLPattern = "https://www.google.com/search?q=%s&lr=lang_en"
+
+// Call handles making a service call to scrape some data from google search engine with a given keyword.
+// It will return an error when the collector can not visit the url.
+func (service *SearchKeywordService) Call() {
+	collector := colly.NewCollector(colly.Async(true))
+	visitURL := fmt.Sprintf(collectingURLPattern, url.QueryEscape(service.Keyword))
+	searchResult := searchKeywordResult{Keyword: service.Keyword, VisitURL: visitURL}
+
+	collector.OnResponse(onResponseHandler)
+	collector.OnRequest(onRequestHandler)
+	collector.OnError(onResponseErrorHandler)
+
+	collector.OnHTML(selectorList["nonAds"], func(element *colly.HTMLElement) {
+		searchResult.NonAds = append(searchResult.NonAds, element.Attr("href"))
+	})
+	collector.OnHTML(selectorList["bottomLinkAds"], func(element *colly.HTMLElement) {
+		searchResult.OtherAds = append(searchResult.OtherAds, element.Attr("href"))
+	})
+	collector.OnHTML(selectorList["otherAds"], func(element *colly.HTMLElement) {
+		searchResult.OtherAds = append(searchResult.OtherAds, element.Attr("href"))
+	})
+	collector.OnHTML(selectorList["topImageAds"], func(element *colly.HTMLElement) {
+		searchResult.TopAds = append(searchResult.TopAds, element.Attr("href"))
+	})
+	collector.OnHTML(selectorList["topLinkAds"], func(element *colly.HTMLElement) {
+		searchResult.TopAds = append(searchResult.TopAds, element.Attr("href"))
+	})
+
+	collector.OnScraped(func(response *colly.Response) {
+		logs.Info(fmt.Sprintf("Search keyword result: %+v", searchResult))
+
+		service.searchResult = &searchResult
+	})
+
+	err := collector.Visit(visitURL)
+	if err != nil {
+		logs.Critical(fmt.Sprintf("Collector visit failed: %v", err))
+	}
+
+	// Disable asynchronous when synchronous flag is enabled
+	if service.synchronous {
```

Comment:
This block will use in test [here](https://github.com/Lahphim/go-crawler-challenge/pull/54/files#diff-913f7310de1279354ddaf7a6dfbecd1ed54c53adccf35587ad2a3ff74c191de0R27)

Because we make this service as [async](https://github.com/Lahphim/go-crawler-challenge/pull/54/files#diff-785dbf8fdffdc312acc4c01ffb3fdb5ccb271f5a55388076e884211f0bc2943fR39), and in some test scenarios I want to get the result from this so I have to disable it with flag.